File: core/planner/src/main/scala/org/apache/calcite/plan/PelagoRelOptCluster.java
package org.apache.calcite.plan;

import ch.epfl.dias.calcite.adapter.pelago.metadata.PelagoRelMetadataProvider;
import org.apache.calcite.plan.RelOptCluster;
import org.apache.calcite.plan.RelOptPlanner;
import org.apache.calcite.plan.RelOptRule;
import org.apache.calcite.plan.RelOptRuleCall;
import org.apache.calcite.rel.RelNode;
import org.apache.calcite.rel.metadata.ChainedRelMetadataProvider;
import org.apache.calcite.rel.metadata.RelMetadataProvider;
import org.apache.calcite.rel.metadata.RelMetadataQuery;
import org.apache.calcite.rel.type.RelDataTypeFactory;
import org.apache.calcite.rex.RexBuilder;

import ch.epfl.dias.calcite.adapter.pelago.metadata.PelagoRelMetadataQuery;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

public class PelagoRelOptCluster extends RelOptCluster {
  PelagoRelOptCluster(final RelOptPlanner planner,
      final RelDataTypeFactory typeFactory,
      final RexBuilder rexBuilder, final AtomicInteger nextCorrel,
      final Map<String, RelNode> mapCorrelToRel) {
    super(planner, typeFactory, rexBuilder, nextCorrel, mapCorrelToRel);
    super.setMetadataProvider(PelagoRelMetadataProvider.INSTANCE);
    setMetadataQuerySupplier(() -> {
      super.setMetadataProvider(ChainedRelMetadataProvider.of(List.of(getMetadataProvider(),
          PelagoRelMetadataProvider.INSTANCE)));
      return PelagoRelMetadataQuery.instance();
    });
  }

  public static RelOptCluster create(RelOptPlanner planner,
                                     RexBuilder rexBuilder) {
    return new PelagoRelOptCluster(planner, rexBuilder.getTypeFactory(),
        rexBuilder, new AtomicInteger(0), new HashMap<String, RelNode>());
  }
}

File: core/planner/src/main/scala/org/apache/calcite/plan/PelagoRelOptCluster.scala
/*
    Proteus -- High-performance query processing on heterogeneous hardware.

                            Copyright (c) 2017
        Data Intensive Applications and Systems Laboratory (DIAS)
                École Polytechnique Fédérale de Lausanne

                            All Rights Reserved.

    Permission to use, copy, modify and distribute this software and
    its documentation is hereby granted, provided that both the
    copyright notice and this permission notice appear in all copies of
    the software, derivative works or modified versions, and any
    portions thereof, and that both notices appear in supporting
    documentation.

    This code is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. THE AUTHORS
    DISCLAIM ANY LIABILITY OF ANY KIND FOR ANY DAMAGES WHATSOEVER
    RESULTING FROM THE USE OF THIS SOFTWARE.
 */

package org.apache.calcite.plan

import ch.epfl.dias.calcite.adapter.pelago.metadata.{
  PelagoRelMetadataProvider,
  PelagoRelMetadataQuery
}
import org.apache.calcite.rel.RelNode
import org.apache.calcite.rel.`type`.RelDataTypeFactory
import org.apache.calcite.rel.metadata.ChainedRelMetadataProvider
import org.apache.calcite.rex.RexBuilder

import java.util
import java.util.concurrent.atomic.AtomicInteger

object PelagoRelOptCluster {
  def create(planner: RelOptPlanner, rexBuilder: RexBuilder) =
    new PelagoRelOptCluster(
      planner,
      rexBuilder.getTypeFactory,
      rexBuilder,
      new AtomicInteger(0),
      new util.HashMap[String, RelNode]
    )
}

class PelagoRelOptCluster private[plan] (
    val planner: RelOptPlanner,
    val typeFactory: RelDataTypeFactory,
    val rexBuilder: RexBuilder,
    val nextCorrel: AtomicInteger,
    val mapCorrelToRel: util.Map[String, RelNode]
) extends RelOptCluster(
      planner,
      typeFactory,
      rexBuilder,
      nextCorrel,
      mapCorrelToRel
    ) {
  super.setMetadataProvider(PelagoRelMetadataProvider.INSTANCE)
  setMetadataQuerySupplier(() => {
    super.setMetadataProvider(
      ChainedRelMetadataProvider
        .of(
          util.List.of(
            getMetadataProvider,
            PelagoRelMetadataProvider.INSTANCE
          )
        )
    )
    PelagoRelMetadataQuery.instance
  })
}
